cdap-gateway/src/main/java/co/cask/cdap/gateway/router/handlers/HttpIdleStateHandler.java

```diff
@@ -0,0 +1,302 @@
+    if (ctx.getPipeline().isAttached()) {
+      // channelOpen event has been fired already, which means
+      // this.channelOpen() will not be invoked.
+      // We have to initialize here instead.
+      initialize(ctx);
+    } else {
+      // channelOpen event has not been fired yet.
+      // this.channelOpen() will be invoked and initialization will occur there.
+    }
+  }
+
+  public void afterAdd(ChannelHandlerContext ctx) throws Exception {
+    // NOOP
+  }
+
+  public void beforeRemove(ChannelHandlerContext ctx) throws Exception {
+    destroy(ctx);
+  }
+
+  public void afterRemove(ChannelHandlerContext ctx) throws Exception {
+    // NOOP
+  }
+
+  @Override
+  public void channelOpen(ChannelHandlerContext ctx, ChannelStateEvent e)
+    throws Exception {
+    // This method will be invoked only if this handler was added
+    // before channelOpen event is fired.  If a user adds this handler
+    // after the channelOpen event, initialize() will be called by beforeAdd().
+    initialize(ctx);
+    ctx.sendUpstream(e);
+  }
+
+  @Override
+  public void channelClosed(ChannelHandlerContext ctx, ChannelStateEvent e)
+    throws Exception {
+    destroy(ctx);
+    ctx.sendUpstream(e);
+  }
+
+  @Override
+  public void messageReceived(ChannelHandlerContext ctx, MessageEvent e)
+    throws Exception {
+    Object message = e.getMessage();
+    if (message instanceof HttpResponse) {
+      HttpResponse response = (HttpResponse) message;
+      if (!response.isChunked()) {
+        markRequestEnd(ctx);
+      }
+    } else if (message instanceof HttpChunk) {
+      HttpChunk chunk = (HttpChunk) message;
+
+      if (chunk.isLast()) {
+        markRequestEnd(ctx);
+      }
+    }
+
+    ctx.sendUpstream(e);
+  }
+
+  private void markRequestEnd(ChannelHandlerContext ctx) {
+    State state = (State) ctx.getAttachment();
+    state.lastMessageTime = System.currentTimeMillis();
+    state.requestInProgress = false;
+  }
+
+  private void markRequestInProgress(ChannelHandlerContext ctx) {
+    // TODO: check if already in progress. if so, return.
+    // this helps avoid the call to System#currentTimeMillis,
+    // at the cost of `lastMessageTime` being inaccurate when requestInProgress is true
+
+    State state = (State) ctx.getAttachment();
+    state.lastMessageTime = System.currentTimeMillis();
+    state.requestInProgress = true;
+  }
+
+  @Override
+  public void writeRequested(ChannelHandlerContext ctx, MessageEvent e) throws Exception {
+    Object message = e.getMessage();
+    if (message instanceof HttpRequest || message instanceof HttpChunk) {
+      markRequestInProgress(ctx);
+    }
+    ctx.sendDownstream(e);
+  }
+
+  private void initialize(ChannelHandlerContext ctx) {
+    State state = state(ctx);
+
+    // Avoid the case where destroy() is called before scheduling timeouts.
+    // See: https://github.com/netty/netty/issues/143
+    synchronized (state) {
```

Comment:
According to the issue - https://github.com/netty/netty/issues/143,
the destroy method can be called before initialize method is complete (and done scheduling).
because of that, if the state is already `DESTROYED`, then we don't start the timer.